### PR TITLE
Add Shipping Zones API Endpoints.

### LINKED ISSUE #10954: Shipping Zones REST API Endpoint
Needs to cover;

- Creation
- Editing
- Getting all zones
- Getting a single zone

Zones should return data about the zone, as well as methods which are within the zone. There is no 'shipping method' api right now so this will be an array list of IDs, which will need expanding once this API does exist.

### PR DESCRIPTION
**Pull request by jeffstieler**
Resolves #10954.

Includes:
- `/shipping/zones`
- `/shipping/zones/<id>`
- `/shipping/zones/<id>/locations`
- `/shipping/zones/<zone_id>/methods`
- `/shipping/zones/<zone_id>/methods/<instance_id>`

### CODE REVIEW

**Comment by justinshreve** [comment]
I'll add this to my list to review :).

File: includes/abstracts/abstract-wc-rest-shipping-zones-controller.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			return new WP_Error( 'rest_no_route', __( 'Shipping is disabled.' ), array( 'status' => 404 ) );
+		}
+
+		if ( ! wc_rest_check_manager_permissions( 'settings', 'edit' ) ) {
+			return new WP_Error( 'woocommerce_rest_cannot_update', __( 'Sorry, you cannot update resource.', 'woocommerce' ), array( 'status' => rest_authorization_required_code() ) );
+		}
+
+		return true;
+	}
+}
```
Comment (justinshreve) [inline includes/abstracts/abstract-wc-rest-shipping-zones-controller.php]: Need an additional line break after last bracket.

File: includes/api/class-wc-rest-shipping-zones-controller.php [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+		if ( ! is_null( $request->get_param( 'name' ) ) ) {
+			$zone->set_zone_name( $request->get_param( 'name' ) );
+		}
+
+		if ( ! is_null( $request->get_param( 'order' ) ) ) {
+			$zone->set_zone_order( $request->get_param( 'order' ) );
+		}
+
+		$zone->create();
```
Comment (justinshreve) [inline includes/api/class-wc-rest-shipping-zones-controller.php]: We should have some error checking here to make sure the zone is actually created correctly. `$zone->get_id()` should !== 0.

**Comment by justinshreve** [comment]
I didn't see anything major -- I can fix up this small stuff now that you are off rotation.

**Comment by justinshreve** [comment]
Shipping zone fixes are in - and the PR looks good to me.

File: includes/abstracts/abstract-wc-rest-shipping-zones-controller.php (hunk 1 above)
Comment (justinshreve) [inline includes/abstracts/abstract-wc-rest-shipping-zones-controller.php]: Fixed

File: includes/api/class-wc-rest-shipping-zones-controller.php [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+		register_rest_route( $this->namespace, '/' . $this->rest_base, array(
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_items' ),
+				'permission_callback' => array( $this, 'get_items_permissions_check' ),
+			),
+			array(
+				'methods'             => WP_REST_Server::CREATABLE,
+				'callback'            => array( $this, 'create_item' ),
+				'permission_callback' => array( $this, 'update_items_permissions_check' ),
```
Comment (justinshreve) [inline includes/api/class-wc-rest-shipping-zones-controller.php]: Just found another one: This should have it's own `create_item_permissions_check` (errors are different).

File: includes/api/class-wc-rest-shipping-zones-controller.php [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+		if ( ! is_null( $request->get_param( 'order' ) ) ) {
+			$zone->set_zone_order( $request->get_param( 'order' ) );
+		}
+
+		$zone->create();
+
+		if ( $zone->get_id() !== 0 ) {
+			$request->set_param( 'id', $zone->get_id() );
+			return $this->get_item( $request );
```
Comment (justinshreve) [inline includes/api/class-wc-rest-shipping-zones-controller.php]: Create should redirect like the other wp-api endpoints.

**Comment by justinshreve** [comment]
Found a few more things today, so needs some more fixes (which I'll handle).

**Comment by mikejolley** [comment]
@justinshreve changing label to in-progress (assuming you're working on the stuff you mentioned above)

**Comment by justinshreve** [comment]
Ok - handled the other feedback I left.

File: includes/abstracts/abstract-wc-rest-shipping-zones-controller.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+	/**
+	 * Check whether a given request has permission to read Shipping Zones.
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|boolean
+	 */
+	public function get_items_permissions_check( $request ) {
+		if ( ! wc_shipping_enabled() ) {
+			return new WP_Error( 'rest_no_route', __( 'Shipping is disabled.' ), array( 'status' => 404 ) );
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-rest-shipping-zones-controller.php]: Missing text domain

File: includes/abstracts/abstract-wc-rest-shipping-zones-controller.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+	/**
+	 * Check if a given request has access to create Shipping Zones.
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|boolean
+	 */
+	public function create_item_permissions_check( $request ) {
+		if ( ! wc_shipping_enabled() ) {
+			return new WP_Error( 'rest_no_route', __( 'Shipping is disabled.' ), array( 'status' => 404 ) );
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-rest-shipping-zones-controller.php]: textdomain

File: includes/abstracts/abstract-wc-rest-shipping-zones-controller.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+	/**
+	 * Check whether a given request has permission to edit Shipping Zones.
+	 *
+	 * @param  WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|boolean
+	 */
+	public function update_items_permissions_check( $request ) {
+		if ( ! wc_shipping_enabled() ) {
+			return new WP_Error( 'rest_no_route', __( 'Shipping is disabled.' ), array( 'status' => 404 ) );
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-rest-shipping-zones-controller.php]: text domain

**Pull request merged**